pkg/controller/seed-controller-manager/default-application-controller/controller.go

```diff
@@ -0,0 +1,442 @@
+	// We don't care about the state of the condition here since it `exists` is enough information to know that the initial application installations controller has already reconciled this cluster.
+	if exists {
+		ignoreDefaultApplications = true
+	}
+
+	// Default applications are already created.
+	if cluster.Status.HasConditionValue(kubermaticv1.ClusterConditionDefaultApplicationInstallationsControllerCreatedSuccessfully, corev1.ConditionTrue) {
+		ignoreDefaultApplications = true
+	}
+
+	// Ensure that cluster is in a state when creating ApplicationInstallation is permissible
+	if !cluster.Status.ExtendedHealth.ApplicationControllerHealthy() {
+		r.log.Debug("Application controller not healthy")
+		return nil, nil
+	}
+
+	// List all ApplicationDefinitions
+	applicationDefinitions := &appskubermaticv1.ApplicationDefinitionList{}
+	if err := r.List(ctx, applicationDefinitions); err != nil {
+		return nil, fmt.Errorf("failed to list ApplicationDefinitions: %w", err)
+	}
+
+	// Collect all applications that need to be installed/updated.
+	applications := []appskubermaticv1.ApplicationInstallation{}
+	for _, applicationDefinition := range applicationDefinitions.Items {
+		if applicationDefinition.DeletionTimestamp != nil {
+			continue
+		}
+
+		// Check if the ApplicationDefinition is targeted to the current cluster's datacenter.
+		if val, ok := applicationDefinition.Annotations[appskubermaticv1.ApplicationTargetDatacentersAnnotation]; ok {
+			// Split the list and check if the cluster's datacenter is included.
+			datacenters := strings.Split(val, ",")
+			if !slices.Contains(datacenters, cluster.Spec.Cloud.DatacenterName) {
+				// Skip this ApplicationDefinition if the cluster's datacenter is not in the list
+				continue
+			}
+		}
+		if applicationDefinition.Annotations[appskubermaticv1.ApplicationEnforceAnnotation] == AnnotationTrueValue || (applicationDefinition.Annotations[appskubermaticv1.ApplicationDefaultAnnotation] == AnnotationTrueValue && !ignoreDefaultApplications) {
+			applications = append(applications, r.generateApplicationInstallation(applicationDefinition))
+		}
+	}
+
+	// We don't want to fail the reconciliation if one application fails so we collect all the errors and return them as a single error.
+	var errors []error
+	for _, application := range applications {
+		// Using reconciler framework here doesn't help since the namespaces are different for the application installations.
+		err := r.ensureApplicationInstallation(ctx, application, cluster)
+		if err != nil {
+			errors = append(errors, err)
+		}
+	}
+
+	if len(errors) == 0 && !cluster.Status.HasConditionValue(kubermaticv1.ClusterConditionDefaultApplicationInstallationsControllerCreatedSuccessfully, corev1.ConditionTrue) {
+		if err := kubermaticv1helper.UpdateClusterStatus(ctx, r.Client, cluster, func(c *kubermaticv1.Cluster) {
+			c.Status.Encryption = nil
```

Comment:
I think this shouldn't be here.

Reply:
Wow, that could have been some fuck up. Thanks for pointing it out.